src/Workspaces/Core/Portable/Shared/Extensions/SyntaxTreeExtensions.cs

```diff
@@ -72,6 +72,46 @@ public static bool IsScript(this SyntaxTree syntaxTree)
             return default;
         }
 
+        /// <summary>
+        /// Returns the first Token that passes <paramref name="predicate"/> within supplied <paramref name="span"/>.
+        /// </summary>
+        public static async Task<SyntaxToken> GetLeftmostTokenInSpanAsync(
+            this SyntaxTree syntaxTree,
+            TextSpan span, 
+            Predicate<SyntaxToken> predicate,
+            CancellationToken cancellationToken,
+            bool findInsideTrivia = false,
+            bool includeSkipped = true,
+            bool includeDirectives = false,
+            bool includeDocumentationComments = false)
+        {
+            Contract.ThrowIfNull(syntaxTree);
+
+            if (span.Start >= syntaxTree.Length)
+            {
+                return default;
+            }
+
+            var root = await syntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(span.Start, findInsideTrivia);
+
+            while (!predicate(token))
+            {
+                if (token.FullSpan.End >= syntaxTree.Length)
+                {
+                    // SyntaxKind = None
+                    return default;
+                }
+
+                token = root.FindTokenOnRightOfPosition(
```

Comment:
is it guaranteed that token will always move to next one?

Reply:
If I understood docs correctly Tokens' `FullSpan` should cover the whole document and thus the location of one's `FullSpan.End` should always be beginning of another's `FullSpan.Start`. If these assumptions are true then I think it is guaranteed (unless `token.GetNextToken` and its docs aren't up to date).

Reply:
Isn't necessary in the end.